aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1344 @@
+    ///     - `Ok(true)` if transitioned from Executing to Executed
+    ///     - `Ok(false)` if transitioned from Aborted to PendingScheduling
+    /// - `Err` if the current state doesn't allow finishing execution
+    pub(crate) fn finish_execution(
+        &self,
+        finished_incarnation: Incarnation,
+    ) -> Result<bool, PanicError> {
+        // TODO(BlockSMTv2): Handle waiting workers when supported (defer waking up).
+
+        let inner_status = &mut *self.inner_status.lock();
+
+        // An incarnation of a transaction can only increase when both finish_execution and
+        // try_abort take effect for the prior incarnation. However, finish_execution is
+        // invoked once per incarnation, and thus incarnations must always match.
+        if inner_status.incarnation() != finished_incarnation {
+            return Err(code_invariant_error(format!(
+                "Finish execution of incarnation {}, but inner status {:?}",
+                finished_incarnation, inner_status,
+            )));
+        }
+
+        match inner_status.status {
+            StatusEnum::Executing => {
+                inner_status.status = StatusEnum::Executed;
+
+                let new_flag = if self.num_stalls.load(Ordering::Relaxed) == 0 {
+                    DependencyFlag::Safe
+                } else {
+                    DependencyFlag::Defer
+                };
+                self.swap_dependency_flag_any(
+                    &[DependencyFlag::Executing],
+                    new_flag,
+                    "finish_execution",
+                )?;
+
+                Ok(true)
+            },
+            StatusEnum::Aborted => {
+                self.incarnate(inner_status, finished_incarnation + 1, false);
+                Ok(false)
+            },
+            StatusEnum::PendingScheduling | StatusEnum::Executed => {
+                Err(code_invariant_error(format!(
+                    "Status update to Executed failed, previous inner status {:?}",
+                    inner_status
+                )))
+            },
+        }
+    }
+
+    /// Attempts to mark a transaction incarnation for abort.
+    ///
+    /// This is the first step of the two-step abort process. It serves as an efficient
+    /// test-and-set filter for abort attempts, ensuring only one worker successfully
+    /// initiates the abort of a specific incarnation.
+    ///
+    /// # Parameters
+    /// - `incarnation`: The incarnation number to abort
+    ///
+    /// # Returns
+    /// - `Ok(true)` if this call successfully marked the incarnation for abort
+    /// - `Ok(false)` if the incarnation was already marked for abort
+    /// - `Err` if the provided incarnation is invalid (greater than the current value)
+    pub(crate) fn start_abort(&self, incarnation: Incarnation) -> Result<bool, PanicError> {
+        let prev_value = self
+            .next_incarnation_to_abort
+            .fetch_max(incarnation + 1, Ordering::Relaxed);
```

Comment:
Can you explain why relaxed here?

Reply:
So what's going on here is the following:
- consider all fetch-x instructions on the same variable. Since fetching and successful x (whatever update) must be atomic, all successful invocations will actually be ordered properly, because of the atomicity of fetch-AND-x, i.e. if you did a lot of fetch-and-increments, even with relaxed ordering, fetches would observe 0,1,2,3, etc. Otherwise, you would have f&i in between some other f and corresponding increment, breaking atomicity.
- Now for unsuccessful operations and already aborted checks, basically note that the counter is monotonically increasing and that those checks are optimization / best effort.